### PR TITLE
Player name entry improvements

### PR DESCRIPTION
**Pull request by Terin**
## Feature Details
This PR makes a couple of adjustments for the name entry function in the conversation panel. First, both text boxes will be visible at once, one with a slightly darker color, so it is apparent you will need to make two entries at the end of the intro conversation, and moving between them looks better. Second, it provides a flicker functionality, where either pressing enter with one completed text box, or entering an invalid character, will cause the selected text box to jump between two colors for a moment, letting the player know something is not correct.

## UI Video
https://user-images.githubusercontent.com/4471575/107439223-588c7b00-6af7-11eb-9ad3-24cfca22ce4b.mp4

## Testing Done
Entered text in the first name field and attempted to hit enter multiple times, then vice versa. Also attempted each forbidden character, and verify they flickered as well.

### CODE REVIEW

File: data/interfaces.txt [hunk 1]
```diff
@@ -18,6 +18,7 @@ color "bright" .75 .75 .75 0.
 color "medium" .5 .5 .5 0.
 color "dim" .25 .25 .25 0.
 color "faint" .1 .1 .1 0.
+color "extrafaint" .05 .05 .05 0.
```
Comment (Amazinite) [inline data/interfaces.txt:21]: EIther two words "extra faint" (no other color definition smashes the words together like this), or choose a different word. How about this?
```suggestion
color "dark" .05 .05 .05 0.
```

File: source/ConversationPanel.cpp [hunk 2]
```diff
@@ -112,6 +112,7 @@ void ConversationPanel::Draw()
 	const Color &dim = *GameData::Colors().Get("dim");
 	const Color &grey = *GameData::Colors().Get("medium");
 	const Color &bright = *GameData::Colors().Get("bright");
+	const Color &extrafaint = *GameData::Colors().Get("extrafaint");
```
Comment (Amazinite) [inline source/ConversationPanel.cpp:115]:
```suggestion
	const Color &dark = *GameData::Colors().Get("dark");
```

File: source/ConversationPanel.cpp [hunk 3]
```diff
@@ -142,15 +143,23 @@ void ConversationPanel::Draw()
...
-			FillShader::Fill(center, fieldSize, selectionColor);
+			// Color selected text box, or flicker if user attempts an error
+			if(flickerTime % 6 > 3)
+				FillShader::Fill(center, fieldSize, dim);
+			else
+				FillShader::Fill(center, fieldSize, selectionColor);
+			if(flickerTime)
+				--flickerTime;
+			// Fill non-selected text box with dimmer color
+			FillShader::Fill(unselected, fieldSize, extrafaint);
```
Comment (Amazinite) [inline source/ConversationPanel.cpp:162]:
```suggestion
			FillShader::Fill(unselected, fieldSize, dark);
```

File: source/ConversationPanel.cpp [hunk 4]
```diff
@@ -142,15 +143,23 @@ void ConversationPanel::Draw()
...
 				continue;
 			}
 			
-			// Fill in whichever entry box is active right now.
-			FillShader::Fill(center, fieldSize, selectionColor);
+			// Color selected text box, or flicker if user attempts an error
+			if(flickerTime % 6 > 3)
+				FillShader::Fill(center, fieldSize, dim);
+			else
+				FillShader::Fill(center, fieldSize, selectionColor);
```
Comment (Amazinite) [inline source/ConversationPanel.cpp:158]: I'd simplify this with a ternary operator.
```suggestion
			FillShader::Fill(center, fieldSize, (flickerTime % 6 > 3) ? dim : selectionColor);
```

**Review by Amazinite — CHANGES_REQUESTED** [review]

File: data/interfaces.txt (hunk 1 above)
Comment (Terin) [reply data/interfaces.txt:21]: Hah, damn it. Was wracking my brain for a word darker than faint, and it was right there all along.

File: source/ConversationPanel.h [hunk 5]
```diff
@@ -112,6 +112,8 @@ template <class T>
 	// The current choices being presented to you:
 	std::list<Paragraph> choices;
 	int choice;
+	// Flicker time, set if the player enters invald input for a pilot's name
```
Comment (Amazinite) [inline source/ConversationPanel.h:115]:
```suggestion
	// Flicker time, set if the player enters invalid input for a pilot's name.
```

File: source/ConversationPanel.cpp [hunk 6]
```diff
@@ -142,15 +143,20 @@ void ConversationPanel::Draw()
...
 			// Handle mouse clicks in whatever field is not selected.
 			if(side != choice)
 			{
 				AddZone(Rectangle(center, fieldSize), [this, side](){ this->ClickName(side); });
 				continue;
 			}
 			
-			// Fill in whichever entry box is active right now.
-			FillShader::Fill(center, fieldSize, selectionColor);
+			// Color selected text box, or flicker if user attempts an error
```
Comment (Amazinite) [inline source/ConversationPanel.cpp:154]:
```suggestion
			// Color selected text box, or flicker if user attempts an error.
```

File: source/ConversationPanel.cpp [hunk 7]
```diff
@@ -142,15 +143,20 @@ void ConversationPanel::Draw()
...
+			FillShader::Fill(center, fieldSize, (flickerTime % 6 > 3) ? dim : selectionColor);
+			if(flickerTime)
+				--flickerTime;
+			// Fill non-selected text box with dimmer color
```
Comment (Amazinite) [inline source/ConversationPanel.cpp:158]:
```suggestion
			// Fill non-selected text box with dimmer color.
```

**Review by Amazinite — CHANGES_REQUESTED** [review]

**Review by tehhowch — COMMENTED** [review]
Rather than give the ConversationPanel even more responsibility in an area that's not really its job, let's extract the "(re)name the player" functionality into a new Panel whose sole job is to draw 2 "validated" inputs at a given position and then pass their values to a callback on exit.

In theory we wouldn't need a whole new panel class for this, but we don't have an intermediate abstraction like an `InputZone` that wraps a `ClickZone` with keypress handling.

**Comment by Terin** [comment]
Are there any others examples of UI classes that call children which scroll with them? I've been trying chip away at this in a few ways, but creating a whole new class was outside the scope of this PR when I made it.

**Comment by tehhowch** [comment]
> children which scroll with them?

You can either disable scrolling in the new panel, or bind a const reference to the provided draw position anchor (which should be a class member variable in the host panel and not a temporary) and let the host panel handle scrolling by not trapping all events and returning false in the scroll handle.

E.g.
```c++
class NewPanel : public Panel {
public:
  NewPanel(const Point &anchor, Validator v)
    : drawAnchor(anchor), validationFn(v)
  {}
  // ...
  bool Scroll(...) override { return false; }
  bool TrapAllEvents() const override { return false; }
  // ...
private:
  const Point &drawAnchor;
  Validation validationFn;
  // ...
};
```
The associated interfaces.txt definition would then be combined with the draw anchor to position the input(s) appropriately.

**Comment by Terin** [comment]
Re-opening after talking with @quyykk / @warp-core, where even they weren't sure how to proceed with refactoring this code into a new class. When/if we wanted to do that factor, the additional code in this PR is simple enough to include in that, so I'm now looking to get this merged.

**Review by quyykk — APPROVED** [review]
Even though it's not great that the conversation panel does yet another thing, I'm not sure how to do this with a panel.

> In theory we wouldn't need a whole new panel class for this, but we don't have an intermediate abstraction like an InputZone that wraps a ClickZone with keypress handling.

this sounds much better, but can always be done later

**Review by warp-core — APPROVED** [review]

**Pull request merged**